File: clients/pascal/Stream.pas

Unit Stream;

{$mode delphi}{$H+}

Interface

Type 
  TByteArray = Array Of Byte;
  TStream = Class
    Public 
      Function ReadBytesAtMost(byteCount: Integer): TByteArray;
      virtual;
      abstract;

      Procedure Write(bytes: TByteArray);
      virtual;
      abstract;

      Procedure Flush;
      virtual;
      abstract;

      Function ReadBytes(byteCount: Integer): TByteArray;
      Function ReadBoolean: Boolean;
      Function ReadInt32: Int32;
      Function ReadInt64: Int64;
      Function ReadSingle: Single;
      Function ReadDouble: Double;
      Function ReadString: String;

      Procedure WriteBoolean(value: Boolean);
      Procedure WriteInt32(value: Int32);
      Procedure WriteInt64(value: Int64);
      Procedure WriteSingle(value: Single);
      Procedure WriteDouble(value: Double);
      Procedure WriteString(value: String);
  End;

Implementation

Uses SysUtils;

Type 
  TInt32Repr = array[0..3] Of Byte;
  TInt64Repr = array[0..7] Of Byte;
  TSingleRepr = array[0..3] Of Byte;
  TDoubleRepr = array[0..7] Of Byte;

Function TStream.ReadBytes(byteCount: Integer): TByteArray;

Var 
  buffer: TByteArray;
  start, i: Integer;
Begin
  result := TByteArray.Create;
  While Length(result) <> byteCount Do
    Begin
      start := Length(result);
      buffer := ReadBytesAtMost(byteCount - Length(result));
      If Length(buffer) = 0 Then
        raise Exception.Create('Unexpected EOF');
      SetLength(result, Length(result) + Length(buffer));
      For i := 0 To Length(buffer) - 1 Do
        result[start + i] := buffer[i];
    End;
End;

Function TStream.ReadBoolean: Boolean;
Begin
  result := ReadBytes(1)[0] <> 0;
End;

Function TStream.ReadInt32: Int32;

Var 
  repr: TInt32Repr;
  bytes: TByteArray;
  i: Integer;
Begin
  bytes := ReadBytes(Length(repr));
  For i := 0 To High(repr) Do
    repr[i] := bytes[i];
  result := LEtoN(Int32(repr));
End;

Function TStream.ReadInt64: Int64;

Var 
  repr: TInt64Repr;
  bytes: TByteArray;
  i: Integer;
Begin
  bytes := ReadBytes(Length(repr));
  For i := 0 To High(repr) Do
    repr[i] := bytes[i];
  result := LEtoN(Int64(repr));
End;

Function TStream.ReadSingle: Single;

Var 
  repr: TSingleRepr;
  bytes: TByteArray;
  i: Integer;
Begin
  bytes := ReadBytes(Length(repr));
  For i := 0 To High(repr) Do
    repr[i] := bytes[i];
  result := Single(repr);
End;

Function TStream.ReadDouble: Double;

Var 
  repr: TDoubleRepr;
  bytes: TByteArray;
  i: Integer;
Begin
  bytes := ReadBytes(Length(repr));
  For i := 0 To High(repr) Do
    repr[i] := bytes[i];
  result := Double(repr);
End;

Function TStream.ReadString: String;
Begin
  result := String(TEncoding.UTF8.GetString(ReadBytes(ReadInt32)));
End;

Procedure TStream.WriteBoolean(value: Boolean);
Begin
  Write(TByteArray.Create(ord(value)));
End;

Procedure TStream.WriteInt32(value: Int32);

Var 
  repr: TInt32Repr;
  bytes: TByteArray;
  i: Integer;

Begin
  repr := TInt32Repr(NtoLE(value));
  SetLength(bytes, Length(repr));
  For i := 0 To High(repr) Do
    bytes[i] := repr[i];
  Write(bytes);
  Finalize(bytes);
End;

Procedure TStream.WriteInt64(value: Int64);

Var 
  repr: TInt64Repr;
  bytes: TByteArray;
  i: Integer;

Begin
  repr := TInt64Repr(NtoLE(value));
  SetLength(bytes, Length(repr));
  For i := 0 To High(repr) Do
    bytes[i] := repr[i];
  Write(bytes);
  Finalize(bytes);
End;

Procedure TStream.WriteSingle(value: Single);

Var 
  repr: TSingleRepr;
  bytes: TByteArray;
  i: Integer;

Begin
  repr := TSingleRepr(value);
  SetLength(bytes, Length(repr));
  For i := 0 To High(repr) Do
    bytes[i] := repr[i];
  Write(bytes);
  Finalize(bytes);
End;

Procedure TStream.WriteDouble(value: Double);

Var 
  repr: TDoubleRepr;
  bytes: TByteArray;
  i: Integer;

Begin
  repr := TDoubleRepr(value);
  SetLength(bytes, Length(repr));
  For i := 0 To High(repr) Do
    bytes[i] := repr[i];
  Write(bytes);
  Finalize(bytes);
End;

Procedure TStream.WriteString(value: String);

Var 
  bytes: TByteArray;
Begin
  bytes := TEncoding.UTF8.GetBytes(UnicodeString(value));
  WriteInt32(Length(bytes));
  Write(bytes);
End;

End.

File: clients/php/Stream.php
<?php

define('LITTLE_ENDIAN', pack('L', 1) === pack('V', 1));
assert(strlen(pack('g', 0.0)) == 4);
assert(strlen(pack('e', 0.0)) == 8);

abstract class InputStream
{
    abstract function readAtMost(int $byteCount): string;
    function read(int $byteCount): string
    {
        $result = '';
        while (strlen($result) < $byteCount) {
            $result .= $this->readAtMost($byteCount - strlen($result));
        }
        return $result;
    }
    function readBool(): bool
    {
        $byte = unpack('C', $this->read(1))[1];
        if ($byte == 0) {
            return false;
        } elseif ($byte == 1) {
            return true;
        } else {
            throw new Exception('bool should be 0 or 1');
        }
    }
    function readInt32(): int
    {
        $bytes = $this->read(4);
        if (!LITTLE_ENDIAN) {
            $bytes = strrev($bytes);
        }
        return unpack('l', $bytes)[1];
    }
    function readInt64(): int
    {
        $bytes = $this->read(8);
        if (!LITTLE_ENDIAN) {
            $bytes = strrev($bytes);
        }
        return unpack('q', $bytes)[1];
    }
    function readFloat32(): float
    {
        return unpack('g', $this->read(4))[1];
    }
    function readDouble(): float
    {
        return unpack('e', $this->read(8))[1];
    }
    function readString(): string
    {
        return $this->read($this->readInt32());
    }
}

abstract class OutputStream
{
    abstract function write(string $bytes): void;
    abstract function flush(): void;
    function writeBool(bool $value): void
    {
        $this->write(pack('C', $value ? 1 : 0));
    }
    function writeInt32(int $value): void
    {
        $bytes = pack('l', $value);
        if (!LITTLE_ENDIAN) {
            $bytes = strrev($bytes);
        }
        $this->write($bytes);
    }
    function writeInt64(int $value): void
    {
        $bytes = pack('q', $value);
        if (!LITTLE_ENDIAN) {
            $bytes = strrev($bytes);
        }
        $this->write($bytes);
    }
    function writeFloat32(float $value): void
    {
        $bytes = pack('g', $value);
        if (!LITTLE_ENDIAN) {
            $bytes = strrev($bytes);
        }
        $this->write($bytes);
    }
    function writeDouble(float $value): void
    {
        $bytes = pack('e', $value);
        if (!LITTLE_ENDIAN) {
            $bytes = strrev($bytes);
        }
        $this->write($bytes);
    }
    function writeString(string $value): void
    {
        $this->writeInt32(strlen($value));
        $this->write($value);
    }
}